File: Js_Solutions/binary_heap/min_binary_heap.js
// Implement a min binary heap data structure

class MinBinaryHeap {
    constructor() {
        this.values = []
    }

    insert(val) {
        this.values.push(val);
        this.bubbleUp();
        return this.values;
    }

    bubbleUp() {
        let idx = this.values.length - 1;
        let element = self.values[idx];
        while (idx > 0) {
            let parentIdx = Math.floor((idx - 1) / 2);
            let parent = this.values[parentIdx];
            if (element >= parent) {
                break;
            }
            this.values[idx] = this.values[parentIdx];
            this.values[parentIdx] = element;
            idx = parentIdx;
        }
    }

    extractMin() {
        let min = this.values[0];
        const last = this.values.pop();
        if (this.values.length > 0) {
            this.values[0] = last;
            this.sinkDown();
        }
        return min;
    }

    sinkDown() {
        let idx = 0;
        const element = this.values[idx];
        const length = this.values.length;

        while (true) {
            let leftChildIdx = 2 * idx + 1;
            let rightChildIdx = 2 * idx + 2;
            let leftChild, rightChild;
            let swap = null;

            if (leftChildIdx < length) {
                leftChild = this.values[leftChildIdx];

                if (leftChild < element) {
                    swap = leftChildIdx;
                }
            }

            if (rightChildIdx < length) {
                rightChild = this.values[rightChildIdx];

                if ((swap === null && rightChild < element) || (swap !== null && rightChild < leftChild)) {
                    swap = rightChildIdx;
                }
            }

            if (swap === null) {
                break;
            }

            this.values[idx] = this.values[swap];
            this.values[swap] = element;
            idx = swap;
        }
    }
}

File: Python_Solutions/binary_heap/min_binary_heap.py
# Implement a min binary heap

class MinBinaryHeap:
    def __init__(self):
        self.values = []
        
    def insert(self, val):
        self.values.append(val)
        self.bubbleUp()
        return self.values
    
    def bubbleUp(self):
        idx = len(self.values) -1
        element = self.values[idx]
        while idx > 0:
            parentIdx = (idx -1)//2
            parent = self.values[parentIdx]
            if parent <= element:
                break
            self.values[idx] = self.values[parentIdx]
            self.values[parentIdx] = element
            idx = parentIdx
            
    def extractMin(self):
        min_val = self.values[0]
        last = self.values.pop()
        if len(self.values) > 0:
            self.values[0] = last
            self.sinkDown()
        return min_val
    
    def sinkDown(self):
        idx = self.values[0]
        element = self.values[idx]
        length = len(self.values)
        while True:
            leftChildIdx = 2*idx + 1
            rightChildIdx = 2*idx + 2
            leftChild = None
            rightChild = None
            swap = None
            
            if leftChildIdx < length:
                leftChild = self.values[leftChildIdx]
                if leftChild < element:
                    swap = leftChildIdx
                    
            if rightChildIdx < length:
                rightChild = self.values[rightChildIdx]
                if (swap is None and rightChild < element) or (swap is not None and rightChild < leftChild):
                    swap = rightChildIdx
            
            if swap is None:
                break
            self.values[idx] = self.values[swap]
            self.values[swap] = element
            idx = swap